tests/e2e/tests/pilot/ingressgateway_test.go

```diff
@@ -66,6 +70,101 @@ func TestIngressGateway(t *testing.T) {
 	})
 }
 
+func TestIngressGateway503DuringRuleChange(t *testing.T) {
+	if !tc.V1alpha3 {
+		t.Skipf("Skipping %s: v1alpha3=false", t.Name())
+	}
+
+	istioNamespace := tc.Kube.IstioSystemNamespace()
+	ingressGatewayServiceName := tc.Kube.IstioIngressGatewayService()
+
+	gateway := &deployableConfig{
+		Namespace: tc.Kube.Namespace,
+		YamlFiles: []string{"testdata/v1alpha3/ingressgateway.yaml"},
+	}
+
+	// Add subsets
+	newDestRule := &deployableConfig{
+		Namespace: tc.Kube.Namespace,
+		YamlFiles: []string{"testdata/v1alpha3/rule-503test-destinationrule-c.yaml"},
+	}
+
+	// route to subsets
+	newVirtService := &deployableConfig{
+		Namespace: tc.Kube.Namespace,
+		YamlFiles: []string{"testdata/v1alpha3/rule-503test-virtualservice.yaml"},
+	}
+
+	addMoreSubsets := &deployableConfig{
+		Namespace: tc.Kube.Namespace,
+		YamlFiles: []string{"testdata/v1alpha3/rule-503test-destinationrule-c-add-subset.yaml"},
+	}
+
+	routeToNewSubsets := &deployableConfig{
+		Namespace: tc.Kube.Namespace,
+		YamlFiles: []string{"testdata/v1alpha3/rule-503test-update-virtualservice.yaml"},
+	}
+
+	deleteOldSubsets := &deployableConfig{
+		Namespace: tc.Kube.Namespace,
+		YamlFiles: []string{"testdata/v1alpha3/rule-503test-destinationrule-c-del-subset.yaml"},
+	}
+
+	if err := gateway.Setup(); err != nil {
+		t.Fatal(err)
+	}
+	defer gateway.Teardown()
+
+	// Update configs in background
+	go func() {
+		log.Infof("Adding new subsets v1,v2")
+		// these functions have built in sleep. So we don't have to add any extra sleep here
+		if err := newDestRule.Setup(); err != nil {
+			t.Fatal(err)
+		}
+		defer newDestRule.Teardown()
+		log.Infof("routing to v1,v2")
+		if err := newVirtService.Setup(); err != nil {
+			t.Fatal(err)
+		}
+		defer newVirtService.Teardown()
+		log.Infof("Adding new subsets v3,v4")
+		if err := addMoreSubsets.Setup(); err != nil {
+			t.Fatal(err)
+		}
+		log.Infof("routing to v3,v4")
+		if err := routeToNewSubsets.Setup(); err != nil {
+			t.Fatal(err)
+		}
+		log.Infof("deleting old subsets v1,v2")
+		if err := deleteOldSubsets.Setup(); err != nil {
+			t.Fatal(err)
+		}
+	}()
```

Comment:
I assume you verified that all of this is threadsafe? :)

Reply:
?? I was hoping that the go people who wrote all this framework stuff would have ensured this. If thats not the case let me know where to look